### ISSUE TITLE
Track Java source-to-class mapping more precisely

### ISSUE THREAD

**Issue by rschiu** [gradle/gradle#2996]
When incremental compilation is enabled, there can be issue if the inner class is in a separate source file.

This issue could arise when annotation processor is used to generate source files.  Most likely due to the use of Kotlin, where kapt is separate from Java's apt.

In the project, there are two source files, HelloWorld.java and HelloWorld$$InnerClass.java.  Running a clean compile will create a both HelloWorld.class and HelloWorld$$InnerClass.class.  Modify HelloWorld.java and run compile again and only HelloWorld.class is created.

### Expected Behavior
Both HelloWorld.class and HelloWorld$$InnerClass.class will be generated.

### Current Behavior
Only HelloWorld.class is generated.  HelloWorld$$InnerClass.class is missing.

### Context
This was found to be an issue in the android plugin using Kotlin. https://issuetracker.google.com/65519025

### Steps to Reproduce (for bugs)
1. Use the attached Gradle project, which is just a simple Java project with incremental compilation enabled.
2. Run ./gradlew clean compileJava
3. You should find 3 .class files being produced
$ find . -name *.class
./build/classes/java/main/com/example/Foo.class
./build/classes/java/main/com/example/HelloWorld$$InnerClass.class
./build/classes/java/main/com/example/HelloWorld.class

4. Edit HelloWorld.java
5. Run ./gradlew compileJava (without clean)
6. Build will be successful, but HelloWorld$$InnerClass.class is missing.

$ find . -name *.class
./build/classes/java/main/com/example/Foo.class
./build/classes/java/main/com/example/HelloWorld.class

[innerClassIssue.zip](https://github.com/gradle/gradle/files/1323197/innerClassIssue.zip)

**Comment by oehme** [comment]
This is not an inner class. Inner classes must be defined in the same source file. It just so happens that the produced class file has the same name as an inner class would have. This can be fixed in at least two ways:

1. An annotation processor should **never** create classes with such names, because they will clash with **real** inner classes written by the user.
2. If source locations are attached to the bytecode, Gradle could figure out precisely which target classes to delete. At the moment we use a heuristic by saying "If it starts with ClassName$$", it was generated from "ClassName.java".

The first one would be a much simpler and quicker fix for now.

**Comment by rschiu** [comment]
The key issue is that the incremental compilation and clean build is different.  If we decide to go with 1., then the code should have never compiled in the first place.
This is from our user, so I am not sure how feasible it is for them to modify the annotation processor.

**Comment by dlew** [comment]
Here's another report of the issue from a slightly different angle: https://issuetracker.google.com/issues/66698564

The issue here is that if there's an inner class that uses annotation processing to generate another class, that generated class gets deleted. For example:

1. Start with inner class "Foo$Bar"
1. Annotation processor generates a new class using the pattern "<ClassName>_Generated".
1. Thus, "Foo$Bar_Generated" is generated.
1. Edit "Foo"
1. On next incremental build, "Foo$Bar_Generated" is missing.

In other words, it's not the annotation processor's intention to create a class name with a $ in it.

**Comment by oehme** [comment]
Incremental compilation is turned off automatically when annotation processors are used. So either processor detection is broken or the class is somehow generated **before** java compilation.

**Comment by rschiu** [comment]
Yes, this is annotation processor for Kotlin, which was run before java compilation.  So is the recommendation for the Kotlin plugin to manually turn off incremental compilation when kapt is used?

**Comment by oehme** [comment]
I'm a bit confused, since the example above shows a Java class. Does Kotlin run Java annotation processing? 

Does this still happen with the latest Kotlin plugin version?

Turning off Java incremental compile in that case is a feasible workaround until this issue is fixed.

**Comment by rschiu** [comment]
Yes, this happens with the latest Kotlin plugin.  Kotlin is running the Java annotation processor because the processor is specified through the kapt configuration from the 'kotlin-kapt' plugin.

**Comment by oehme** [comment]
Got it, thanks for the explanation!

To fix this, Gradle would need to ensure that it only deletes a class if it was really compiled from a stale source file. Right now we only do [simple pattern matching](https://github.com/gradle/gradle/blob/c86b62b42603d0797e29d4566eef322a37add795/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalCompilationInitializer.java#L61-L61). 

One approach would be to read the source information from the .class file as part of the incremental dependency analysis. The incremental compiler could then use this information to know exactly which class file was created from which source file. 

If you or someone else would like to give this a shot, we'll be happy to provide some more code pointers.

**Comment by oehme** [comment]
As discussed on our call, here is another, simpler solution:

For every `.class` file we delete, look if there is a `.java` file of the same name and add that to the recompilation as well.

**Comment by rschiu** [comment]
It seems the easiest solution is to just change the sourceToCompile pattern to match the classesToDelete pattern.  WDYT?

**Comment by oehme** [comment]
Good idea!

**Comment by oehme** [comment]
This was fixed by #3133
